filters/filter.go

```diff
@@ -90,29 +111,82 @@ func newFilter(pattern string) (Filter, error) {
 	}
 
 	if wIdx == len(pattern)-1 {
-		// Wildcard at end
-		return newStartsWithFilter(pattern[:len(pattern)-1]), nil
+		// Single wildcard at end
+		return newRangeFilter(pattern[:len(pattern)-1], false, start)
 	}
 
 	secondWIdx := strings.Index(pattern[wIdx+1:], wildcardChar)
 	if secondWIdx == -1 {
 		if wIdx == 0 {
-			return newEndsWithFilter(pattern[1:]), nil
+			// Single wildcard at start
+			return newRangeFilter(pattern[1:], true, end)
 		}
 
-		return NewMultiFilter([]Filter{
-			newStartsWithFilter(pattern[:wIdx]),
-			newEndsWithFilter(pattern[wIdx+1:]),
-		}, Conjunction), nil
+		// Single wildcard in the middle
+		first, err := newRangeFilter(pattern[:wIdx], false, start)
+		if err != nil {
+			return nil, err
+		}
+
+		second, err := newRangeFilter(pattern[wIdx+1:], true, end)
+		if err != nil {
+			return nil, err
+		}
+
+		return NewMultiFilter([]Filter{first, second}, Conjunction), nil
 	}
 
 	if wIdx == 0 && secondWIdx == len(pattern)-2 && len(pattern) > 2 {
-		return newContainsFilter(pattern[1 : len(pattern)-1]), nil
+		// Wildcard at beginning and end
+		return newContainsFilter(pattern[1 : len(pattern)-1])
 	}
 
 	return nil, errInvalidFilterPattern
 }
 
+// newRangeFilter creates a filter that checks for ranges (? or [] or {}) and segments
+// the pattern into a multiple chain filters based on ranges found
+func newRangeFilter(pattern string, backwards bool, seg chainSegment) (Filter, error) {
+	var filters []chainFilter
+	eqIdx := -1
+	for i := 0; i < len(pattern); i++ {
+		if pattern[i] == singleRangeStartChar {
+			if eqIdx != -1 {
+				filters = append(filters, newEqualityChainFilter(pattern[eqIdx:i], backwards))
+				eqIdx = -1
+			}
+
+			endIdx := strings.Index(pattern[i:], singleRangeEndStr)
```

Comment:
I can't use pattern[i+1:] at this point as i might be len(pattern), in which case we'll get an out of bounds. I can have an additional if statement for that case I guess, but it makes the code more messy.